desc/basis.py

```diff
@@ -1414,9 +1413,584 @@ def zernike_radial_poly(r, l, m, dr=0, exact="auto"):
+    r_jacobi = 1 - 2 * r**2
+    m = jnp.abs(m)
+    n = ((l - m) // 2).astype(int)
+
+    M_max = jnp.max(m)
+    # Loop over every different m value. There is another nested
+    # loop which will execute necessary n values.
+    out = fori_loop(0, (M_max + 1).astype(int), body, (out))
+
+    return out
+
+
+def jacobi_poly_single(x, n, alpha, beta=0, P_n1=0, P_n2=0):
+    """Evaluate Jacobi for single alpha and n pair."""
+    c = 2 * n + alpha + beta
+    a1 = 2 * n * (c - n) * (c - 2)
+    a2 = (c - 1) * (c * (c - 2) * x + (alpha - beta) * (alpha + beta))
+    a3 = 2 * (n + alpha - 1) * (n + beta - 1) * c
+
+    a1 = jnp.where(a1 == 0, 1e-6, a1)
+    P_n = (a2 * P_n1 - a3 * P_n2) / a1
+    # Checks for special cases
+    P_n = jnp.where(n < 0, 0, P_n)
+    P_n = jnp.where(n == 0, 1, P_n)
+    P_n = jnp.where(n == 1, (alpha + 1) + (alpha + beta + 2) * (x - 1) / 2, P_n)
+    return P_n
+
+
+@custom_jvp
+@jit
 def zernike_radial(r, l, m, dr=0):
     """Radial part of zernike polynomials.
 
+    Calculates Radial part of Zernike Polynomials using Jacobi recursion relation
+    by getting rid of the redundant calculations for appropriate modes. This version
+    is almost the same as zernike_radial_old function but way faster and more
+    accurate. First version of this function is zernike_radial_separate which has
+    many function for each derivative definition. User can refer that for clarity.
+
+    There was even faster version of this code but that doesn't have checks
+    for duplicate modes
+
+    # Find the index corresponding to the original array
+    # I changed arange function to get rid of 0 as index confusion
+    # so if index is full of 0s, there is no such mode
+    # (FAST BUT NEED A CHECK FOR DUPLICATE MODES)
+    index = jnp.where(
+        jnp.logical_and(m == alpha, n == N),
+        jnp.arange(1, m.size + 1),
+        0,
+    )
+    idx = jnp.sum(index)
+    # needed for proper index
+    idx -= 1
+    result = (-1) ** N * r**alpha * P_n
+    out = out.at[:, idx].set(jnp.where(idx >= 0, result, out.at[:, idx].get()))
+
+    Above part replaces the matrix update conducted by following code,
+
+    _, _, _, out = fori_loop(0, m.size, update, (alpha, N, result, out))
+
+    Parameters
+    ----------
+    r : ndarray, shape(...,)
+        radial coordinates to evaluate basis
+    l : ndarray of int, shape(K,)
+        radial mode number(s)
+    m : ndarray of int, shape(K,)
+        azimuthal mode number(s)
+    dr : int
+        order of derivative (Default = 0)
+
+    Returns
+    -------
+    out : ndarray, shape(...,K)
+        basis function(s) evaluated at specified points
+
+    """
+    return _zernike_radial_vectorized(r, l, m, dr)
+
+
+@functools.partial(jnp.vectorize, excluded=(1, 2, 3), signature="()->(k)")
+def _zernike_radial_vectorized(r, l, m, dr):
+    # need this to be separate so that we can have a default value for dr and not
+    # vectorize over it.
+    def update(i, args):
```

Comment:
Can you add some comments describing what the different parts of this function are doing